base/linalg/lq.jl

```diff
@@ -0,0 +1,170 @@
+end
+
+copy(A::LQ) = LQ(copy(A.factors), copy(A.τ))
+convert{T}(::Type{LQ{T}},A::LQ) = LQ(convert(AbstractMatrix{T}, A.factors), convert(Vector{T}, A.τ))
+convert{T}(::Type{Factorization{T}}, A::LQ) = convert(LQ{T}, A)
+ctranspose{T}(A::LQ{T}) = QR{T,typeof(A.factors)}(A.factors', A.τ)
+
+function getindex(A::LQ, d::Symbol)
+    m, n = size(A)
+    if d == :L
+        return tril!(A.factors[1:m, 1:min(m,n)])
+    elseif d == :Q
+        return LQPackedQ(A.factors,A.τ)
+    else
+        throw(KeyError(d))
+    end
+end
+
+function getindex(A::LQPackedQ, i::Integer, j::Integer)
+    x = zeros(eltype(A), size(A, 1))
+    x[i] = 1
+    y = zeros(eltype(A), size(A, 2))
+    y[j] = 1
+    return dot(x, A*y)
+end
+
+getq(A::LQ) = LQPackedQ(A.factors, A.tau)
+
+convert{T}(::Type{LQPackedQ{T}}, Q::LQPackedQ) = LQPackedQ(convert(AbstractMatrix{T}, Q.factors), convert(Vector{T}, Q.τ))
+convert{T}(::Type{AbstractMatrix{T}}, Q::LQPackedQ) = convert(LQPackedQ{T}, Q)
+
+size(A::LQ, dim::Integer) = size(A.factors, dim)
+size(A::LQ) = size(A.factors)
+size(A::LQPackedQ, dim::Integer) = 0 < dim ? (dim <= 2 ? size(A.factors, dim) : 1) : throw(BoundsError())
+size(A::LQPackedQ) = size(A.factors)
+
+full(A::LQ) = A[:L]*A[:Q]
+#=
+We construct the full eye here, even though it seems ineffecient, because
+every element in the output matrix is a function of all the elements of
+the input matrix. The eye is modified by the elementary reflectors held
+in A, so this is not just an indexing operation. Note that in general
+explicitly constructing Q, rather than using the ldiv or mult methods,
+may be a wasteful allocation.
+=#
+function full{T}(A::LQPackedQ{T}; thin::Bool=true)
+    if thin
+        LAPACK.orglq!(copy(A.factors),A.τ)
+    else
+        A_mul_B!(A, eye(T, size(A.factors,2), size(A.factors,1)))
+    end
+end
+
+## Multiplication by LQ
+A_mul_B!{T<:BlasFloat}(A::LQ{T}, B::StridedVecOrMat{T}) = A[:L]*LAPACK.ormlq!('L','N',A.factors,A.τ,B)
+A_mul_B!{T<:BlasFloat}(A::LQ{T}, B::QR{T}) = L*LAPACK.ormlq!('L','N',A.factors,A.τ,full(B))
+A_mul_B!{T<:BlasFloat}(A::QR{T}, B::LQ{T}) = A_mul_B!(zeros(full(A)), full(A), full(B))
+function *{TA,TB}(A::LQ{TA},B::StridedVecOrMat{TB})
+    TAB = promote_type(TA, TB)
+    A_mul_B!(convert(Factorization{TAB}, A), TB==TAB ? copy(B) : copy_oftype(B, TAB))
+end
+function *{TA,TB}(A::LQ{TA},B::QR{TB})
+    TAB = promote_type(TA, TB)
+    A_mul_B!(convert(Factorization{TAB}, A), TB==TAB ? copy(B) : convert(Factorization{TAB}, B))
+end
+function *{TA,TB}(A::QR{TA},B::LQ{TB})
+    TAB = promote_type(TA, TB)
+    A_mul_B!(convert(Factorization{TAB}, A), TB==TAB ? copy(B) : convert(Factorization{TAB}, B))
+end
+
+## Multiplication by Q
+### QB
+A_mul_B!{T<:BlasFloat}(A::LQPackedQ{T}, B::StridedVecOrMat{T})   = LAPACK.ormlq!('L','N',A.factors,A.τ,B)
+function *{TA,TB}(A::LQPackedQ{TA},B::StridedVecOrMat{TB})
+    TAB = promote_type(TA, TB)
+    A_mul_B!(convert(AbstractMatrix{TAB}, A), TB==TAB ? copy(B) : copy_oftype(B, TAB))
```

Comment:
I was too fast there. No need to change the promotion of `A`, but for `B` it should just be `copy_oftype(B, TAB)` instead of `TB==TAB ? copy(B) : copy_oftype(B, TAB)`